### PR TITLE
More mainline fixes

### PR DESCRIPTION
**Pull request by sklam**
continue on #7000

### CODE REVIEW

**Comment by sklam** [comment]
BFID: `numba_smoketest_cpu_yaml_26`

**Comment by stuartarchibald** [comment]
> BFID: `numba_smoketest_cpu_yaml_26`

Passed.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_52`.

**Comment by stuartarchibald** [comment]
This looks stable on both CPU and CUDA hardware.

File: numba/core/utils.py [hunk 1]
```diff
@@ -296,26 +296,66 @@ def __setitem__(self, key, value):
...
-            return self
-        res = instance.__dict__[self.name] = self.func(instance)
-        return res
+if PYVERSION > (3, 7):
+    from functools import cached_property
+else:
+    from threading import RLock
+
+    # The following cached_property() implementation is adapted from CPython:
+    # https://github.com/python/cpython/blob/3.8/Lib/functools.py#L924-L976
```
Comment (stuartarchibald) [inline numba/core/utils.py:305]: Explicit SHA?

File: numba/core/overload_glue.py [hunk 2]
```diff
@@ -183,16 +199,36 @@ def __call__(self, func, typing_key=None):
...
 def glue_typing(concrete_function, typing_key=None):
     """This is a decorator for wrapping the typing part for a concrete function
     'concrete_function', it's a text-only replacement for '@infer_global'"""
     return _overload_glue(concrete_function,
                           typing_key=typing_key).wrap_typing()
 
 
+_deferred_lowering = defaultdict(list)
+
+
```
Comment (stuartarchibald) [inline numba/core/overload_glue.py:214]: Am wondering if having these in globals is going to lead to race conditions later?

**Review by stuartarchibald — COMMENTED** [review]
@sklam @gmarkall given this patch is a group effort, I've just review the bits which are not from me!

File: numba/core/utils.py (hunk 1 above)
Comment (sklam) [reply numba/core/utils.py:305]: added in https://github.com/numba/numba/pull/7003/commits/15f9a1f28ce6f3d78ce9716166cae221c2c0dd38

Comment (stuartarchibald) [reply numba/core/utils.py:305]: thanks

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Review by sklam — COMMENTED** [review]
I have reviewed and approving changes  b12814a 4453540 2cdad85

**Pull request merged**